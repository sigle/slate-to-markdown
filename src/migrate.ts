import { Text } from 'slate';
/**
 * Inspired by https://gist.github.com/webel/545f229fe79c2176dbaed9023de46e12
 *
 * Slate's schema has changed vastly under 2 years. The text editor is still
 * a better candidate than the other OSS editors out there, so we must live
 * with the major changes.
 *
 * Migrate a schema from the old version 0.33 to current version 0.6x
 * Inspiration taken wholly from
 * https://github.com/react-page/react-page/blob/b6c83a8650cfe9089e0c3eaf471ab58a0f7db761/packages/plugins/content/slate/src/migrations/v004.ts
 */

const migrateTextNode = (oldNode: any) => {
  const leaves = {
    text: oldNode.text,
    ...oldNode.marks?.reduce(
      (acc: any, mark: any) => ({
        ...acc,
        [mark.type]: true,
      }),
      {}
    ),
  };
  return leaves;
};

const migrateLinkNode = (node: any) => {
  return {
    link: node.data.href,
    type: node.type,
    children: node.nodes?.map(migrateNode).flat() ?? [],
  };
};

const migrateImageNode = (node: any) => {
  return {
    link: node.data.src ? encodeURI(node.data.src) : '',
    caption: '',
    type: node.type,
    children: [],
  };
};

// List items needs to be wrapped in a paragraph
const migrateListItemNode = (node: any) => {
  return {
    data: node.data ?? {},
    type: 'list_item',
    children: [
      {
        type: 'paragraph',
        children: node.nodes?.map(migrateNode).flat() ?? [],
      },
    ],
  };
};

const migrateElementNode = (node: any) => {
  return {
    data: node.data ?? {},
    type: node.type,
    children: node.nodes?.map(migrateNode).flat() ?? [],
  };
};

const migrateNode = (oldNode: any) => {
  if (oldNode.object === 'text') {
    return migrateTextNode(oldNode);
  } else if (oldNode.object === 'inline' && oldNode.type === 'link') {
    return migrateLinkNode(oldNode);
  } else if (oldNode.object === 'block' && oldNode.type === 'image') {
    return migrateImageNode(oldNode);
  } else if (oldNode.object === 'block' && oldNode.type === 'list-item') {
    return migrateListItemNode(oldNode);
  } else {
    return migrateElementNode(oldNode);
  }
};

/**
 * Migration of data < slate 0.46, remove the leaves property which has been
 * removed in 0.46.
 * For performance, only call Slate if the leaves property is detected.
 */
const removeLeaves = (oldNode: any) => {
  let newNodes: any[] = [];
  oldNode.nodes.forEach((node: any) => {
    // Inline and block can contain an array of node that needs to be migrated
    if ((node.object === 'inline' || node.object === 'block') && node.nodes) {
      newNodes.push(removeLeaves(node));
      return;
    }

    if (node.object === 'text' && node.leaves) {
      newNodes = newNodes.concat(
        Text.createList(node.leaves)
          .toArray()
          .map((data) => data.toJSON())
      );
      return;
    }

    newNodes.push(node);
  });

  oldNode.nodes = newNodes;
  return oldNode;
};

/**
 * Remove the new line char that can be in the titles as it can break the markdown.
 */
const fixNewLineTitle = (oldNode: any) => {
  if (
    oldNode.object === 'block' &&
    (oldNode.type === 'heading-one' ||
      oldNode.type === 'heading-two' ||
      oldNode.type === 'heading-three')
  ) {
    oldNode.nodes = oldNode.nodes.map((node: any) =>
      node.leaves
        ? {
            ...node,
            leaves: node.leaves.map((leaf: any) =>
              leaf.text.replace(/^\n/, '')
            ),
          }
        : node
    );
  }
  return oldNode;
};

export const migrateSchema = (nodes: any) => {
  nodes = nodes.map(fixNewLineTitle);
  return nodes.map(removeLeaves).map(migrateNode);
};
