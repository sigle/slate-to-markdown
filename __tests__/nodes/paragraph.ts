import { convert } from '../../src';

it('render text', () => {
  const slateJSON = [
    {
      object: 'block',
      type: 'paragraph',
      data: {},
      nodes: [
        {
          object: 'text',
          text: 'Hello',
          marks: [],
        },
      ],
    },
  ];
  expect(convert(slateJSON)).toMatchSnapshot();
});

it('render bold text', () => {
  const slateJSON = [
    {
      object: 'block',
      type: 'paragraph',
      data: {},
      nodes: [
        {
          object: 'text',
          text: 'Hello ',
          marks: [],
        },
        {
          object: 'text',
          text: 'bold',
          marks: [
            {
              object: 'mark',
              type: 'bold',
              data: {},
            },
          ],
        },
        {
          object: 'text',
          text: ' hello',
          marks: [],
        },
      ],
    },
  ];
  expect(convert(slateJSON)).toMatchSnapshot();
});

it('render italic text', () => {
  const slateJSON = [
    {
      object: 'block',
      type: 'paragraph',
      data: {},
      nodes: [
        {
          object: 'text',
          text: 'Hello ',
          marks: [],
        },
        {
          object: 'text',
          text: 'italic',
          marks: [
            {
              object: 'mark',
              type: 'italic',
              data: {},
            },
          ],
        },
        {
          object: 'text',
          text: ' hello',
          marks: [],
        },
      ],
    },
  ];
  expect(convert(slateJSON)).toMatchSnapshot();
});

it('render underlined text', () => {
  const slateJSON = [
    {
      object: 'block',
      type: 'paragraph',
      data: {},
      nodes: [
        {
          object: 'text',
          text: 'Hello ',
          marks: [],
        },
        {
          object: 'text',
          text: 'underlined',
          marks: [
            {
              object: 'mark',
              type: 'underlined',
              data: {},
            },
          ],
        },
        {
          object: 'text',
          text: ' hello',
          marks: [],
        },
      ],
    },
  ];
  expect(convert(slateJSON)).toMatchSnapshot();
});

it('render all marks text', () => {
  const slateJSON = [
    {
      object: 'block',
      type: 'paragraph',
      data: {},
      nodes: [
        {
          object: 'text',
          text: 'Hello ',
          marks: [],
        },
        {
          object: 'text',
          text: 'bold, italic, underlined',
          marks: [
            {
              object: 'mark',
              type: 'bold',
              data: {},
            },
            {
              object: 'mark',
              type: 'italic',
              data: {},
            },
            {
              object: 'mark',
              type: 'underlined',
              data: {},
            },
          ],
        },
        {
          object: 'text',
          text: ' hello',
          marks: [],
        },
      ],
    },
  ];
  expect(convert(slateJSON)).toMatchSnapshot();
});

it('render code text', () => {
  const slateJSON = [
    {
      object: 'block',
      type: 'paragraph',
      data: {},
      nodes: [
        {
          object: 'text',
          text: 'Hello ',
          marks: [],
        },
        {
          object: 'text',
          text: 'code',
          marks: [
            {
              object: 'mark',
              type: 'code',
              data: {},
            },
          ],
        },
        {
          object: 'text',
          text: ' hello',
          marks: [],
        },
      ],
    },
  ];
  expect(convert(slateJSON)).toMatchSnapshot();
});

/**
 * slate < 0.46
 */

it('render slate < 0.46 text', () => {
  const slateJSON = [
    {
      object: 'block',
      type: 'paragraph',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: 'Hello',
              marks: [],
            },
          ],
        },
      ],
    },
  ];
  expect(convert(slateJSON)).toMatchSnapshot();
});

it('merge slate < 0.46 text with mark', () => {
  const slateJSON = [
    {
      object: 'block',
      type: 'paragraph',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: 'Of course, you can edit your text at any moment and make it ',
              marks: [],
            },
            {
              object: 'leaf',
              text: 'bold',
              marks: [
                {
                  object: 'mark',
                  type: 'bold',
                  data: {},
                },
              ],
            },
            {
              object: 'leaf',
              text: ', ',
              marks: [],
            },
            {
              object: 'leaf',
              text: 'italic',
              marks: [
                {
                  object: 'mark',
                  type: 'italic',
                  data: {},
                },
              ],
            },
            {
              object: 'leaf',
              text: ' or ',
              marks: [],
            },
            {
              object: 'leaf',
              text: 'underlined',
              marks: [
                {
                  object: 'mark',
                  type: 'underlined',
                  data: {},
                },
              ],
            },
            {
              object: 'leaf',
              text: '.',
              marks: [],
            },
          ],
        },
      ],
    },
  ];
  expect(convert(slateJSON)).toMatchSnapshot();
});

it('render slate < 0.46 without crashing', () => {
  const slateJSON = [
    {
      object: 'block',
      type: 'paragraph',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: 'This guide provides you with everything you need to know about the text editor and page styles. It may change, depending on the added themes or possible updates.',
              marks: [],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'heading-one',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: 'Images',
              marks: [],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'heading-two',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: 'Story cover picture',
              marks: [],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'paragraph',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: 'You can add a cover picture to any stories.\nBy clicking on the wheel next to the save button on the "Edit" page, you will open the settings section.\nIn this column, you will be able to add a cover picture.',
              marks: [],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'image',
      data: {
        src: 'https://i.goopics.net/kwlWG.jpg',
      },
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: '',
              marks: [],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'paragraph',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: 'This is a full width image.\nFor now, we recommend you use an image host such ',
              marks: [],
            },
          ],
        },
        {
          object: 'inline',
          type: 'link',
          data: {
            href: 'https://goopics.net/',
          },
          nodes: [
            {
              object: 'text',
              leaves: [
                {
                  object: 'leaf',
                  text: 'Goopics',
                  marks: [],
                },
              ],
            },
          ],
        },
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: ".\nDon't worry, ",
              marks: [],
            },
            {
              object: 'leaf',
              text: 'we will add a picture uploader very soon.',
              marks: [
                {
                  object: 'mark',
                  type: 'bold',
                  data: {},
                },
              ],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'paragraph',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: '',
              marks: [],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'heading-two',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: 'Image gallery',
              marks: [],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'paragraph',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: 'Coming soon...',
              marks: [
                {
                  object: 'mark',
                  type: 'italic',
                  data: {},
                },
              ],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'paragraph',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: '',
              marks: [
                {
                  object: 'mark',
                  type: 'italic',
                  data: {},
                },
              ],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'heading-one',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: 'This is an H1',
              marks: [],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'paragraph',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: 'Quisque facilisis erat a dui. Nam malesuada ornare dolor. Cras gravida, diam sit amet rhoncus ornare, erat elit consectetuer erat, id egestas pede nibh eget odio. Proin tincidunt, velit vel porta elementum, magna diam molestie sapien, non aliquet massa pede eu diam. Aliquam iaculis. Fusce et ipsum et nulla tristique facilisis.',
              marks: [],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'heading-two',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: '\nThis is an H2',
              marks: [],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'paragraph',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: 'Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Morbi commodo, ipsum sed pharetra gravida, orci magna rhoncus neque, id pulvinar odio lorem non turpis. Nullam sit amet enim. Suspendisse id velit vitae ligula volutpat condimentum. Aliquam erat volutpat. Sed quis velit. Nulla facilisi. Nulla libero.',
              marks: [],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'paragraph',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: '',
              marks: [],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'heading-three',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: 'This is an H3',
              marks: [],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'paragraph',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: 'Aliquam sodales neque at diam efficitur, eu venenatis ipsum dapibus. Maecenas id condimentum ligula. Fusce purus odio, malesuada tincidunt sapien vel, fringilla pulvinar eros. Class aptent taciti sociosqu ad litora torquent per conubia nostra, per inceptos himenaeos. Aenean ipsum felis, aliquam ut purus id, cursus laoreet sem.',
              marks: [],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'paragraph',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: '',
              marks: [],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'heading-one',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: 'Drop cap',
              marks: [],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'paragraph',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: 'Coming soon...',
              marks: [
                {
                  object: 'mark',
                  type: 'italic',
                  data: {},
                },
              ],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'paragraph',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: '',
              marks: [
                {
                  object: 'mark',
                  type: 'italic',
                  data: {},
                },
              ],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'heading-one',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: 'Quoting',
              marks: [],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'block-quote',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: 'Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Morbi commodo, ipsum sed pharetra gravida, orci magna rhoncus neque, id pulvinar odio lorem non turpis. Nullam sit amet enim. Suspendisse id velit vitae ligula volutpat condimentum. Aliquam erat volutpat. Sed quis velit. Nulla facilisi. Nulla libero.',
              marks: [],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'paragraph',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: '',
              marks: [],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'heading-one',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: 'Code Blocks',
              marks: [],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'paragraph',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: 'Coming soon...',
              marks: [
                {
                  object: 'mark',
                  type: 'italic',
                  data: {},
                },
              ],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'paragraph',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: 'You will be able to customize your theme by adding code lines in this section.\nAvailable very soon.',
              marks: [],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'paragraph',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: '',
              marks: [],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'heading-one',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: 'Lists',
              marks: [],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'heading-two',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: 'Unordered Lists',
              marks: [],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'paragraph',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: 'You can create unordered lists, typically rendered as a bulleted list.',
              marks: [],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'bulleted-list',
      data: {},
      nodes: [
        {
          object: 'block',
          type: 'list-item',
          data: {},
          nodes: [
            {
              object: 'text',
              leaves: [
                {
                  object: 'leaf',
                  text: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.\nVestibulum luctus elit ut metus feugiat maximus.\nMorbi tempus purus.',
                  marks: [],
                },
              ],
            },
          ],
        },
        {
          object: 'block',
          type: 'list-item',
          data: {},
          nodes: [
            {
              object: 'text',
              leaves: [
                {
                  object: 'leaf',
                  text: 'Integer sagittis ex at odio feugiat lobortis.',
                  marks: [],
                },
              ],
            },
          ],
        },
        {
          object: 'block',
          type: 'list-item',
          data: {},
          nodes: [
            {
              object: 'text',
              leaves: [
                {
                  object: 'leaf',
                  text: 'Ut sit amet quam ac erat sollicitudin laoreet ornare ac neque.\nPraesent dignissim nulla interdum, venenatis mi eget, eleifend urna.',
                  marks: [],
                },
              ],
            },
          ],
        },
        {
          object: 'block',
          type: 'list-item',
          data: {},
          nodes: [
            {
              object: 'text',
              leaves: [
                {
                  object: 'leaf',
                  text: 'Maecenas egestas est eget fringilla mollis.',
                  marks: [],
                },
              ],
            },
          ],
        },
        {
          object: 'block',
          type: 'list-item',
          data: {},
          nodes: [
            {
              object: 'text',
              leaves: [
                {
                  object: 'leaf',
                  text: 'Cras et nisl ut enim pellentesque dignissim.',
                  marks: [],
                },
              ],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'heading-two',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: '\nOrdered Lists',
              marks: [],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'paragraph',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: 'You also can create ordered lists, typically rendered as a numbered list.',
              marks: [],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'numbered-list',
      data: {},
      nodes: [
        {
          object: 'block',
          type: 'list-item',
          data: {},
          nodes: [
            {
              object: 'text',
              leaves: [
                {
                  object: 'leaf',
                  text: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit.\nVestibulum luctus elit ut metus feugiat maximus.\nMorbi tempus purus.',
                  marks: [],
                },
              ],
            },
          ],
        },
        {
          object: 'block',
          type: 'list-item',
          data: {},
          nodes: [
            {
              object: 'text',
              leaves: [
                {
                  object: 'leaf',
                  text: 'Integer sagittis ex at odio feugiat lobortis.',
                  marks: [],
                },
              ],
            },
          ],
        },
        {
          object: 'block',
          type: 'list-item',
          data: {},
          nodes: [
            {
              object: 'text',
              leaves: [
                {
                  object: 'leaf',
                  text: 'Ut sit amet quam ac erat sollicitudin laoreet ornare ac neque.\nPraesent dignissim nulla interdum, venenatis mi eget, eleifend urna.',
                  marks: [],
                },
              ],
            },
          ],
        },
        {
          object: 'block',
          type: 'list-item',
          data: {},
          nodes: [
            {
              object: 'text',
              leaves: [
                {
                  object: 'leaf',
                  text: 'Maecenas egestas est eget fringilla mollis.',
                  marks: [],
                },
              ],
            },
          ],
        },
        {
          object: 'block',
          type: 'list-item',
          data: {},
          nodes: [
            {
              object: 'text',
              leaves: [
                {
                  object: 'leaf',
                  text: 'Cras et nisl ut enim pellentesque dignissim.',
                  marks: [],
                },
              ],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'paragraph',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: '',
              marks: [],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'heading-one',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: 'Text edition',
              marks: [],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'paragraph',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: 'Of course, you can edit your text at any moment and make it ',
              marks: [],
            },
            {
              object: 'leaf',
              text: 'bold',
              marks: [
                {
                  object: 'mark',
                  type: 'bold',
                  data: {},
                },
              ],
            },
            {
              object: 'leaf',
              text: ', ',
              marks: [],
            },
            {
              object: 'leaf',
              text: 'italic',
              marks: [
                {
                  object: 'mark',
                  type: 'italic',
                  data: {},
                },
              ],
            },
            {
              object: 'leaf',
              text: ' or ',
              marks: [],
            },
            {
              object: 'leaf',
              text: 'underlined',
              marks: [
                {
                  object: 'mark',
                  type: 'underlined',
                  data: {},
                },
              ],
            },
            {
              object: 'leaf',
              text: '.',
              marks: [],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'paragraph',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: 'You can use shortcuts for that: ',
              marks: [],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'bulleted-list',
      data: {},
      nodes: [
        {
          object: 'block',
          type: 'list-item',
          data: {},
          nodes: [
            {
              object: 'text',
              leaves: [
                {
                  object: 'leaf',
                  text: 'Bold: ctrl+b (cmd+b on mac)',
                  marks: [],
                },
              ],
            },
          ],
        },
        {
          object: 'block',
          type: 'list-item',
          data: {},
          nodes: [
            {
              object: 'text',
              leaves: [
                {
                  object: 'leaf',
                  text: 'Italic: ctrl+i (cmd+i on mac)',
                  marks: [],
                },
              ],
            },
          ],
        },
        {
          object: 'block',
          type: 'list-item',
          data: {},
          nodes: [
            {
              object: 'text',
              leaves: [
                {
                  object: 'leaf',
                  text: 'Underlined: ctrl+u (cmd+u on mac)',
                  marks: [],
                },
              ],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'paragraph',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: '',
              marks: [],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'heading-two',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: 'Links',
              marks: [],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'paragraph',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: 'Easily add links to your text by highlighting it and clicking on the ',
              marks: [],
            },
          ],
        },
        {
          object: 'inline',
          type: 'link',
          data: {
            href: 'https://app.sigle.io/sigleapp.id.blockstack/L7I4iV6bYQ8WYvuT3RcoM',
          },
          nodes: [
            {
              object: 'text',
              leaves: [
                {
                  object: 'leaf',
                  text: '"link"',
                  marks: [],
                },
              ],
            },
          ],
        },
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: ' button in the text editor.',
              marks: [],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'paragraph',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: '',
              marks: [],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'heading-one',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: 'Videos',
              marks: [],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'paragraph',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: 'Coming soon...',
              marks: [
                {
                  object: 'mark',
                  type: 'italic',
                  data: {},
                },
              ],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'heading-one',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: '\nAudio',
              marks: [],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'paragraph',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: 'Coming soon...',
              marks: [
                {
                  object: 'mark',
                  type: 'italic',
                  data: {},
                },
              ],
            },
          ],
        },
      ],
    },
    {
      object: 'block',
      type: 'paragraph',
      data: {},
      nodes: [
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: "\nDon't forget to share your blog with us by mentioning ",
              marks: [],
            },
          ],
        },
        {
          object: 'inline',
          type: 'link',
          data: {
            href: 'https://twitter.com/sigleapp',
          },
          nodes: [
            {
              object: 'text',
              leaves: [
                {
                  object: 'leaf',
                  text: '@sigleapp',
                  marks: [],
                },
              ],
            },
          ],
        },
        {
          object: 'text',
          leaves: [
            {
              object: 'leaf',
              text: ' on Twitter! 🙂',
              marks: [],
            },
          ],
        },
      ],
    },
  ];
  expect(convert(slateJSON)).toMatchSnapshot();
});
