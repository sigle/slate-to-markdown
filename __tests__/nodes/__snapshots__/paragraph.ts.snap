// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`merge slate < 0.46 text with mark 1`] = `
"Of course, you can edit your text at any moment and make it **bold**, _italic_ or ~~underlined~~.
"
`;

exports[`render all marks text 1`] = `
"Hello ~~***bold, italic, underlined***~~ hello
"
`;

exports[`render bold text 1`] = `
"Hello **bold** hello
"
`;

exports[`render code text 1`] = `
"Hello \`code\` hello
"
`;

exports[`render italic text 1`] = `
"Hello _italic_ hello
"
`;

exports[`render slate < 0.46 text 1`] = `
"Hello
"
`;

exports[`render text 1`] = `
"Hello
"
`;

exports[`render underlined text 1`] = `
"Hello ~~underlined~~ hello
"
`;
